pkg/image/prune/imagepruner.go

```diff
@@ -0,0 +1,736 @@
+func DeletingImagePruneFunc(images client.ImageInterface, streams client.ImageStreamsNamespacer) ImagePruneFunc {
+	return func(image *imageapi.Image, referencedStreams []*imageapi.ImageStream) []error {
+		result := []error{}
+
+		glog.V(4).Infof("Deleting image %q", image.Name)
+		if err := images.Delete(image.Name); err != nil {
+			e := fmt.Errorf("Error deleting image: %v", err)
+			glog.Error(e)
+			result = append(result, e)
+			return result
+		}
+
+		var updateStream func(*imageapi.Image, *imageapi.ImageStream, int) error
+		updateStream = func(image *imageapi.Image, stream *imageapi.ImageStream, retry int) error {
+			glog.V(4).Infof("Checking if stream %s/%s has references to image in status.tags", stream.Namespace, stream.Name)
+			for tag, history := range stream.Status.Tags {
+				glog.V(4).Infof("Checking tag %q", tag)
+				newHistory := imageapi.TagEventList{}
+				for i, tagEvent := range history.Items {
+					glog.V(4).Infof("Checking tag event %d with image %q", i, tagEvent.Image)
+					if tagEvent.Image != image.Name {
+						glog.V(4).Infof("Tag event doesn't match deleting image - keeping")
+						newHistory.Items = append(newHistory.Items, tagEvent)
+					}
+				}
+				stream.Status.Tags[tag] = newHistory
+			}
+
+			glog.V(4).Infof("Updating image stream %s/%s", stream.Namespace, stream.Name)
+			glog.V(5).Infof("Updated stream: %#v", stream)
+			//TODO use the updated ImageStream that's returned and update the entry in the graph
+			//to hopefully avoid having to re-get the stream?
+			if _, err := streams.ImageStreams(stream.Namespace).UpdateStatus(stream); err != nil {
+				if errors.IsConflict(err) && retry > 0 {
+					if stream, err := streams.ImageStreams(stream.Namespace).Get(stream.Name); err == nil {
+						return updateStream(image, stream, retry-1)
+					}
+				}
+				return err
+			}
+			return nil
+		}
+
+		for _, stream := range referencedStreams {
+			if err := updateStream(image, stream, retryCount); err != nil {
+				e := fmt.Errorf("Unable to update image stream status %s/%s: %v", stream.Namespace, stream.Name, err)
+				glog.Error(e)
+				result = append(result, e)
+			}
+		}
+
+		return result
+	}
+}
+
+// DeletingLayerPruneFunc returns a LayerPruneFunc that sends the
+// DeleteLayersRequest to the Docker registry.
+//
+// The request URL is http://registryURL/admin/layers and it is a DELETE
+// request.
+//
+// The body of the request is JSON, and it is a map[string][]string, with each
+// key being a layer, and each value being a list of Docker image repository
+// names referenced by the layer.
+func DeletingLayerPruneFunc(registryClient *http.Client) LayerPruneFunc {
+	return func(registryURL, repoName, layer string) error {
+		glog.V(4).Infof("Pruning registry %q, repo %q, layer %q", registryURL, repoName, layer)
+
+		//TODO https
+		req, err := http.NewRequest("DELETE", fmt.Sprintf("http://%s/admin/%s/layers/%s", registryURL, repoName, layer), nil)
```

Comment:
Why not make the protocol part of something higher level (abstract into a method of a class you invoke)?